server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -349,31 +350,36 @@ static void checkForFailedPluginRemovals(final Path pluginsDirectory) throws IOE
     private static Set<Bundle> findBundles(final Path directory, String type) throws IOException {
         final Set<Bundle> bundles = new HashSet<>();
         for (final Path plugin : findPluginDirs(directory)) {
-            final Bundle bundle = readPluginBundle(bundles, plugin, type);
-            bundles.add(bundle);
+            final Bundle bundle = readPluginBundle(plugin, type);
+            if (bundle.plugin.getType() == PluginType.BOOTSTRAP) {
+                logger.trace("--- skipping bootstrap only [{}] [{}]", type, plugin.toAbsolutePath());
+            } else {
+                if (bundles.add(bundle) == false) {
+                    throw new IllegalStateException("duplicate " + type + ": " + bundle.plugin);
+                }
+                if (type.equals("module") && bundle.plugin.getName().startsWith("test-") && Build.CURRENT.isSnapshot() == false) {
+                    throw new IllegalStateException("external test module [" + plugin.getFileName() + "] found in non-snapshot build");
+                }
+            }
         }
 
+        logger.trace(
+            "findBundles(" + type + ") returning: " + bundles.stream().map(b -> b.plugin.getName()).sorted().collect(Collectors.toList())
```

Comment:
nit: this trace is going to produce a string every time. the message should be wrapped in a lambda?

Reply:
I've wrapped it in a lambda.